lib/mfg.c

```diff
@@ -463,3 +485,140 @@ int switchtec_read_sec_cfg_file(FILE *setting_file,
 
 	return 0;
 }
+
+/**
+ * @brief Set KMSK entry
+ * @param[in]  dev		Switchtec device handle
+ * @param[in]  public_key	Public key
+ * @param[in]  public_key_exp	Public key exponent
+ * @param[in]  signature	Signature
+ * @param[in]  kmsk_entry_data	KMSK entry data
+ * @return 0 on success, error code on failure
+ */
+int switchtec_kmsk_set(struct switchtec_dev *dev,
+		       uint8_t *public_key,
+		       uint32_t public_key_exp,
+		       uint8_t *signature,
+		       uint8_t *kmsk_entry_data)
+{
+	int ret;
+	struct kmsk_cmd1 {
+		uint8_t subcmd;
+		uint8_t reserved[3];
+		uint8_t pub_key[SWITCHTEC_PUB_KEY_LEN];
+		uint32_t pub_key_exponent;
+	} cmd1;
+
+	struct kmsk_cmd2 {
+		uint8_t subcmd;
+		uint8_t reserved[3];
+		uint8_t signature[SWITCHTEC_SIG_LEN];
+	} cmd2;
+
+	struct kmsk_cmd3 {
+		uint8_t subcmd;
+		uint8_t num_entries;
+		uint8_t reserved[2];
+		uint8_t kmsk[SWITCHTEC_KMSK_LEN];
+	} cmd3;
+
+	if (public_key) {
```

Comment:
It kind of seems like this would be better as three functions: set_public_key, set signature and set_kmsmk... They don't really seem related and pushing them all into the same function just makes this more complicated.... At the very list split them into three static functions.

Reply:
This is a function which you can use in two scenarios: 
- when the chip is in UNSECURED mode, only KMSK entry data is need.
- when the chip is in SECURED mode, all parameters are needed.

The raw uint8_t* parameter type is not a good choice. I will change them to structure with defined length.

Reply:
All the more reason to separate it into multiple API functions.... Unless you actually want to check for SECURED mode and ensure that all parameters are set. But that's still a bit ugly IMO.

Reply:
Separating them into 3 functions means user could call them in different sequence, which is not readily supported by firmware. We would like to make sure the data are sent in the sequence as is done in the function. So keeping it in one function is a better option at the moment.

Reply:
Well I'd still separate them into three static functions then. All in one is a bit much.